tools/nsc/main.cpp

```diff
@@ -0,0 +1,155 @@
+		if (argc < 2) {
+			m_logger->log("Insufficient arguments.", ILogger::ELL_ERROR);
+			return false;
+		}
+
+		m_arguments = std::vector<std::string>(argv.begin() + 2, argv.end()); // turn argv into vector for convenience
+		std::string file_to_compile = argv[1];
+
+		if (!m_system->exists(file_to_compile, IFileBase::ECF_READ)) {
+			m_logger->log("Incorrect arguments. Expecting second argument to be filename of the shader intended to compile.", ILogger::ELL_ERROR);
+			return false;
+		}
+		std::string output_filepath = "";
+		for (auto i = 1; i < m_arguments.size(); i++)
+		{
+			if (m_arguments[i] == "-no-nbl-builtins")
+			{
+				// unmount builtins
+				m_logger->log("Unmounting builtins.");
+				m_system->unmountBuiltins();
+				no_nbl_builtins = true;
+				m_arguments.erase(m_arguments.begin() + i - 1);
+				i--;
+			}
+			else if (m_arguments[i] == "-Fo")
+			{
+				// get filepath to save output to
+				if (i + 1 < argc)
+				{
+					output_filepath = argv[i + 1];
+					m_logger->log("Compiled shader code will be saved to " + output_filepath);
+					m_arguments.erase(m_arguments.begin() + i, m_arguments.begin() + i + 1);
+					i--;
+				}
+				else
+				{
+					m_logger->log("Incorrect arguments. Expecting filename after -Fo.", ILogger::ELL_ERROR);
+				}
+			}
+
+		}
+
+#ifndef NBL_EMBED_BUILTIN_RESOURCES
+		if (!no_nbl_builtins) {
+			m_system->unmountBuiltins();
+			no_nbl_builtins = true;
+			m_logger->log("nsc.exe was compiled with builtin resources disabled. Force enabling -no-nbl-builtins.", ILogger::ELL_WARNING);
+		}
+#endif
+		string shader_code = open_shader_file(file_to_compile);
+		auto compilation_result = compile_shader(shader_code, file_to_compile);
+
+		// writie compiled shader to file as bytes
+		if (compilation_result && !output_filepath.empty()) {
+			std::fstream output_file(output_filepath, std::ios::out | std::ios::binary);
+			output_file.write((const char*)compilation_result->getContent()->getPointer(), compilation_result->getContent()->getSize());
+			output_file.close();
+			m_logger->log("Shader compilation successful.");
+		}
+		else {
+			m_logger->log("Shader compilation failed.", ILogger::ELL_ERROR);
+		}
+		return true;
+	}
+
+	void workLoopBody() override {}
+
+	bool keepRunning() override { return false; }
+
+
+private:
+
+	core::smart_refctd_ptr<ICPUShader> compile_shader(std::string& shader_code, std::string_view sourceIdentifier) {
+		constexpr uint32_t WorkgroupSize = 256;
+		constexpr uint32_t WorkgroupCount = 2048;
+		const string WorkgroupSizeAsStr = std::to_string(WorkgroupSize);
+		const IShaderCompiler::SPreprocessorOptions::SMacroDefinition WorkgroupSizeDefine = { "WORKGROUP_SIZE",WorkgroupSizeAsStr };
+
+		smart_refctd_ptr<CHLSLCompiler> hlslcompiler = make_smart_refctd_ptr<CHLSLCompiler>(smart_refctd_ptr(m_system));
+
+		CHLSLCompiler::SOptions options = {};
+		// want as much debug as possible
+		options.stage = asset::IShader::E_SHADER_STAGE::ESS_UNKNOWN;
```

Comment:
use the asset loader to load your input file as ICPUShader, it might be able to deduce the stage for you
(in return you have to check that the content type is  HLSL and not something else)

Reply:
still outstanding